augur/measurements.py

```diff
@@ -0,0 +1,364 @@
+                f"because a '{required_column}' column already exists in collection TSV.",
+                file=sys.stderr,
+            )
+            checks_passed = False
+
+    if not checks_passed:
+        return None
+
+    # Rename user provided columns to expected columns
+    collection_df = collection_df.rename(columns=required_column_map)
+
+    # Make sure the value column is numeric
+    try:
+        collection_df['value'] = pd.to_numeric(collection_df['value'])
+    except ValueError as e:
+        print(f"ERROR: Found a non-numeric measurement value: {e}", file=sys.stderr)
+        return None
+
+    return collection_df
+
+
+def get_collection_groupings(collection, grouping_columns):
+    """
+    Creates the groupings for the provided collection using the provided
+    grouping columns after verifying the columns exist in the collection.
+
+    Parameters
+    ----------
+    collection: pandas.DataFrame
+        The collection used to validate groupings
+    grouping_columns: list[str]
+        List of grouping column names
+
+    Returns
+    -------
+    list[dict] or None
+        The groupings for the collection config or None if all grouping columns are invalid
+    """
+    groupings = [{'key': col} for col in grouping_columns if column_exists(collection, col, "grouping")]
+
+    if not groupings:
+        print("ERROR: Provided grouping columns were invalid for provided collection.", file=sys.stderr)
+        return None
+
+    return groupings
+
+
+def override_config_with_args(config, args):
+    """
+    Overrides values in the config with values of provided command line args.
+
+    Parameters
+    ----------
+    config: dict
+        A collection config
+    args: dict
+        The __dict__ attribute of the parsed arguments from argparse
+    """
+    config_key_args = ['key', 'title', 'filters', 'x_axis_label', 'threshold']
+    display_default_args = ['group_by', 'measurements_display', 'show_overall_mean', 'show_threshold']
+
+    for key_arg in config_key_args:
+        if args.get(key_arg) is not None:
+            config[key_arg] = args[key_arg]
+
+    for default_arg in display_default_args:
+        if args.get(default_arg) is not None:
+            config['display_defaults'] = config.get('display_defaults', {})
+            config['display_defaults'][default_arg] = args[default_arg]
+
+
+def validate_output_json(output_json):
+    """
+    Validate the output JSON against the measurements schema
+
+    Parameters
+    ----------
+    output_json: str
+        Filepath to output JSON
+
+    """
+    print("Validating produced measurements JSON")
+    try:
+        validate_measurements_json(measurements_json=output_json)
+    except ValidateError:
+        print(
+            "ERROR: Validation of output JSON failed. See detailed errors above.",
+            file=sys.stderr,
+        )
+        sys.exit(1)
+
+
+def export_measurements(args):
```

Comment:
I've been mulling over this comment and keep waffling on it...

On one hand, I want to push back on moving all of the functions into the "main" function. I think the functions can exist even if we don't plan on reusing them. I view the functions as self-contained bits of logic that can be easier to understand than having to read through the entire "main" function. 

On the other hand, this would be such a straightforward "main" function that I can see the simplicity of moving everything. Then we can avoid unnecessary raising/excepting within functions. 

All to say, I'm not sure what's the best thing to do here 😅  
I can add a final "overhaul" commit that moves everything into "main" to see what it looks like after addressing all of the other comments.

Reply:
Made the overhaul commit https://github.com/nextstrain/augur/pull/879/commits/d22aeeec1e3fdebf5a758e6331cdb3bd00958d6a. Let me know if this looks better than splitting into functions!

Reply:
Thanks, @joverlee521! d22aeee looks great to me, although I should clarify my original comment.

I do think functions are great for abstracting and encapsulating logic, even when that logic is only used once or never exposed through a public API. For example, [moving augur filter's logic into separate functions with a consistent interface](https://github.com/nextstrain/augur/commit/0710b500ee6f50f090636e9858fd3dc86a327952) made their internal logic of each filter easier to check and test, the calling code clearer to read, and new filters easier to add.

My main concern in the comment above was that the functions still behaved as those they were part of the "main" CLI logic by interacting directly with the operating system through print statements and sys exit calls. Those types of interactions are difficult to handle properly in the calling code. For example, if we decided to handle all exceptions of a specific kind at the top-level of Augur (like @victorlin [recently did for FileNotFoundError exceptions](https://github.com/nextstrain/augur/pull/914/files)), we would like to know our functions will raise these exceptions to be handled downstream like that instead of trying to handle the errors internally.

If you preferred to keep the logic from this PR split out into functions, we would just need those functions to handle errors differently than the main CLI function does by raising exceptions.

Reply:
I'm actually pretty happy with the overhaul. It's more straightforward!